src/aws/flb_aws_util.c

```diff
@@ -927,3 +932,74 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag,
         }
         return NULL;
 }
+
+/*
+ * Replace %f and %L in time_key_format with milliseconds and nanoseconds. And get the
+ * time format output and output length with strftime().
+ */
+size_t create_time_formatter(char **out_buf, const char *time_key_format,
+                             struct flb_time *tms)
+{
+    char millisecond_str[FLB_AWS_MILLISECOND_FORMATTER_LENGTH+1];
+    char nanosecond_str[FLB_AWS_NANOSECOND_FORMATTER_LENGTH+1];
+    char *tmp_parsed_time_str;
+    char *tmp_out_buf;
+    size_t out_size;
+    size_t tmp_parsed_time_str_len;
+    size_t time_key_format_len;
+    struct tm timestamp;
+
+    /*
+     * Guess the max_length needed for tmp_parsed_time_str and tmp_out_buf. The
+     * upper bond is 12*strlen(time_key_format) because the worst scenario will be only
+     * %c in time_key_format, and %c will be tansfer to 24 chars long time_format
+     * by function strftime().
+     */
+    tmp_parsed_time_str_len = 12*strlen(time_key_format);
+    /*
+     * Use tmp_parsed_time_str to buffer time_key_format when replace %f and %F
+     * with milliseconds and nanoseconds.
+     */
+    tmp_parsed_time_str = (char *)flb_calloc(tmp_parsed_time_str_len, sizeof(char));
+    if (!tmp_parsed_time_str) {
+        flb_errno();
+        return -1;
+    }
+
+    tmp_out_buf = (char *)flb_calloc(tmp_parsed_time_str_len, sizeof(char));
+    if (!tmp_out_buf) {
+        flb_errno();
+        flb_free(tmp_parsed_time_str);
+        return -1;
+    }
+
+    snprintf(millisecond_str, FLB_AWS_MILLISECOND_FORMATTER_LENGTH+1, "%" PRIu64, (uint64_t) tms->tm.tv_nsec / 1000000);
+    snprintf(nanosecond_str, FLB_AWS_NANOSECOND_FORMATTER_LENGTH+1, "%" PRIu64, (uint64_t) tms->tm.tv_nsec);
+    time_key_format_len = strlen(time_key_format);
+    /* Replace %f to millisecond, %L to nanosecond in time_key_format. */
+    for (int i = 0; i < time_key_format_len; i++) {
+        if (!strncmp(time_key_format+i, FLB_AWS_MILLISECOND_FORMATTER, 2)) {
+            strncat(tmp_parsed_time_str, millisecond_str, FLB_AWS_MILLISECOND_FORMATTER_LENGTH+1);
+            i += 1;
+        } else if (!strncmp(time_key_format+i, FLB_AWS_NANOSECOND_FORMATTER, 2)) {
+            strncat(tmp_parsed_time_str, nanosecond_str, FLB_AWS_NANOSECOND_FORMATTER_LENGTH+1);
+            i += 1;
+        } else {
+            strncat(tmp_parsed_time_str,time_key_format+i,1);
+        }
+    }
+
+    gmtime_r(&tms->tm.tv_sec, &timestamp);
+    out_size = strftime(tmp_out_buf, tmp_parsed_time_str_len, tmp_parsed_time_str, &timestamp);
+    /* Check whether tmp_parsed_time_str_len is enough for tmp_out_buff */
+    if (out_size == 0) {
+        flb_free(tmp_parsed_time_str);
+        flb_free(tmp_out_buf);
```

Comment:
Errno will not be set. I would skip putting flb_errno() here to avoid potential confusion.
https://man7.org/linux/man-pages/man3/strftime.3.html
> If the length of the result string (including the terminating
       null byte) would exceed max bytes, then strftime() returns 0, and
       the contents of the array are undefined.

> If the output string would exceed max bytes, [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is not set.
       This makes it impossible to distinguish this error case from
       cases where the format string legitimately produces a zero-length
       output string.  POSIX.1-2001 does not specify any [errno](https://man7.org/linux/man-pages/man3/errno.3.html) settings
       for strftime().